Dev christoph

CDStark opened the pull request:
AL as another option for training

Commit:
specified requirements versions and added new config file

Commit:
fixed paths to prediction nifti yaml file and some typos

Commit:
fix argparse to not require additional argument for boolean type options

Commit:
update paths to Nifti Data and Model

Commit:
Create new Config Files for TULIP 1_5T and 3T; change plot to axial

Commit:
Enable save predict image as nii file

Add functionality to predict.py in save_img_mat() to save the image as a
 nii file. The filetype gets specified in config with parameter
 save_predict_img_datatype

Commit:
Edit TULIP config files to save result as nii

Commit:
Bugfix - specifying h5py version

the new version of the h5py caused an error so the old version of the
package gets specified in requirements.txt

Commit:
First setup for Active learning implementation

Create the function sort_by_informativeness() and the accompaning
function uncertainty_sampling() as a first step of the active learning
implementation. The idea is for sort_by_informativeness() to sort the
training-patches by a certain value that represents it's potential to
benefit the model in training.

Commit:
First draft for Active learning in train.py

Commit:
Code calculation of uncertainty values

Correct/finish the function uncertainty_sampling() in active_learning.py
that calculates an uncertainty value for a given prediction-Tensor. It
now first calculates a value for every pixel and then averages over the
entire image to get a value for the image.

Commit:
Data loading for active learning

Establish the first part of loading data from tf record files for
subsequent prediction necessary for active learning. The process is
heavily inspired by pipeline.py and predict.py

Commit:
Redesign AL implementation in train.py

To avoid redundancy in code, redesign the implementation of active
learning in train.py. Build of training pipeline and fitting of the
model are now in the same for loop (it doesn't matter if active
learning is on or of)

Commit:
Determine patch indices in active_learning.py

Commit:
Patch data in active_learning.py

Commit:
Rename variables + introduce predict in al

Rename some variables to be better readable. Turn off random shiftig of
patches in patching function. Add prediction statement (doesn't work at
the moment)

Commit:
Predict bugfix and calculate uncertainty in al

Fix prediction error in active_learning.py by casting the indice list to
 Float32. Add part that calculates an uncertainty value for every
 image.

Commit:
Selection of most useful batches

In active_learning.py add code that selects the patches with the highest
 uncertainty value for training and return these to train.py

Commit:
Regularize indices list before prediction

For some networks the indices-list should be regularized before
prediction. Therefore add this process to prediction in
active_learning.py. The process is inspired/copied form predict.py
predict_image(). To keep the code more readable put this in a extra
function predict()

Commit:
Introduce PatchPool class + predict patch select

Introduce the class PatchPool that contains patching relevant parameters
 and also keeps track of all patches their properties and whether or not
 they were already used in a prediction or should be used in the next
 one. For this implement a second class Patch that represents a single
 patch with relevant information.
 The PatchPool class contains methods get_unused_patches_indices() that
 outputs all patches of an image that haven't been used in training the
 network. Implement this method in query_training_patches() to predict
 only the next possible candidates for every iteration.

 The method select_patches is supposed to pick the n best
 patches for training(not tested yet)

Commit:
Rebuild PatchPool with dict

Rebuild the PatchPool class. Use a dictionary which includes a
dictionary for every image that contains the patches. This way acsess to
 specific patches saved in pool is much easier and efficient. To this
 end implement a method get_pos_key() that creates a key for a given
 patch index. Also rewrite the methods calculate_values(),
 get_unused_patches_indices() and select_patches() accordingly.
 Implement the methods in the code of query_training_patches()
 !The PatchPool generally works however the indices are initialized
 before the actual image is patched, if the indexes change during
 patching an error occurs, this will have to be changed in the following
  commit

Commit:
Relocate the initialisation of patches in pool

Don't initialize the patches in PatchPool self.pool on initialisation of
 the class but instead the first time the values of the batches in
 self.pool get updated. Modify get_unused_patches to return the ideal
 list of indices if this initialisation hasn't occured yet. To keep
 Track of this introduce self.patches_set_up, a list that keeps track of
  for which image the patches have already been initialized.

Commit:
Fix get_pos_key for border patches not on lattice

Commit:
Bugfix patch selection in PatchPool

Commit:
Create get_patches_to_train for PatchPool

Create method that returns the patches selected for training for a
specific image number and edits the pool accordingly

Commit:
Rebuild active learning implementation in train.py

Only create PatchPool Object if active learning is activated, pass the
object to pipeline if al activated. Passing of the object replaces
parameter active_learning as marker in pipeline, that al is wished.

Commit:
Edit pipeline and modified fnc get_patches_data()

Make the modified function get_patches_data and it's usage in pipeline
compatible again. Change order of arguments in pipeline, edit return
statement in the function. Ajust query_training_patches() accordingly

Commit:
Fix get_patches_to_train to output indexes

Commit:
Small changes and revert changes to pad_and_patch

Commit:
Enable return indices only in get_patches_data

To only get the information on how the patches indices will turn out
after patching modify the function. The idea was to use this
functionality to build the patch pool for active learning. For now
however another technique is used.

Commit:
ImagePatches class + revert changes to patch fnct.

Create ImagePatches class and it's methods set_up_patches() and
get_pipeline_patches() that keep track of the patches of an image for AL
Revert changes to get_pad_and_patch method

Commit:
Implement ImagePatches in PatchPool

Rebuild PatchPool class to work with ImagePatches class

Commit:
Change pool structure to use simple list

tf Datasets can only handle tensors. Therfore don't use ImagePatches
oject or Patches object and instead us a simple list as pool of patches.
 Instead of using multiple dictionarys as before (pool, to_train,...)
 add a status value to every index in indices.
 (For the identification of the patch after selection through AL the
 index in the list instead of the patch index is planned to be added
 over in pipeline togethe with the image number)

Commit:
Get patches from pool for pipeline

Create method get_patches_pipeline() for PatchPool that gets a list of
the indices of a certain image (identified by tfRecord path).
Depending on the status of the patches and the mode of PatchPool (build
pipeline for querying or training with AL) the needed patch indices are
chosen from the entire pool.
Implement the function in pipeline.py

Commit:
Get id with patches from pool

Change get_patches_pipeline() in active_learning.py (Method of
PatchPool) to output an id value pait consisting of the image number and
 position of the patch in the pool list for every patch

Commit:
Get id from pool in pipeline and pass to _map()

If a pool object is given as argument to pool, get the indice list and
its corrsponding patch id list from the pool and hand them to the _map()
 function

Commit:
Revert changes to pipeline + new requirement modAL

To prepare for new approach of data pipeline and active learning, revert
changes to pipeline.py
Also add the modAL package that is needed for the new approach

Commit:
Convert TFRecords data to hdf5 file of patches

Read the TFRecords files of the train data, pad and patch it and save
the patches and the accompanying labels as a hdf5 file. The process is
done through the tf pipeline in pipeline.py

Commit:
Save data converted to hdf5 in useful structure

Change the conversion function that converts tfRecord and saves it as
hdf5 file, in a way so that the hdf5 file data structure is useful for
later data reading. The data is now organized in image and label data,
the image data is further divided into training and validation data. The
 ids of the images and accompanying labels are the same

Commit:
Save image data in one list + train & val id lists

Instead of storing the data in the hdf5 file separately in a training
and
validation
group, save them in the same group and create two lists train_ids and
val_ids to accumulate the ids of training and validation data

Commit:
Introduce DataGenerator class

Create DataGenerator class that inherits from keras.utils.Sequence and
enables loading of data from hdf5 file in an efficient manner. The code
is based on the blog post:
https://stanford.edu/~shervine/
blog/keras-how-to-generate-data-on-the-fly#
and will be adjusted according to the situation.

Commit:
Adjust DataGenerator class

Adjust the DataGenerator class to work with the data saved in the hdf5
file according to convert_tr_record_hdf5() and the nature of the
segmentation data (e.g. 3D label images instead of only one class value)

Commit:
Conversion to hdf5 only if necessary & Bugfix

Define tf_records_as_hdf5() function in keras_data_generator.py that
checks if a hdf5 file already exists and only executes the old
convert_tf_records_hdf5() if not, saving a lot of calculations.

Also change DataGenerator class to inherit from tf.keras... instead of
keras... to fix bug

Commit:
Create separate AL train process

Commit:
Create uncertainty measure functions

Create functions in active_learning.py that can calculate a measure of
uncertainty from the probabilitys returned by a prediction for the
different classes. Introduced: uncertainty, margin and entropy
The functions can handle multiple predictions at the same time (e.g.
one for every pixel from a segmentation result)

Commit:
Implement query selection

Complete query_selection function, that selects the n best samples from
all samples by their utility value

Commit:
Create CustomActiveLearner class and methods

Introduce the CustomActiveLearner class inspired by modAL, but designed
to work with DataGenerator class and data saved in a hdf5 file.
Create initialisation function with optional initial data and methods
for fitting data as well as adding new data and the teach and query
methods as interface to the AL loop

Commit:
Create al learning loop + bugfix + docstrings

Create the al learning loop in train.py using teach() and query() of the
CustomActiveLearner class
Fix the query method that previously used the wrong indices for output
Add docstrings in active_learning.py

Commit:
Add possibility of fit kwargs in teach() method

Commit:
Add different reduction methods for al

Add two options on how to condense the classification results of the
entire segmentation to a single value. For now add calculations: taking
mean of entire patch before or after the utility values are calculated
from the probability distribution of the classes

Commit:
Connect config settings with fit (& predict) in AL

Pass relevant config settings to fit method in al loop. The batch size
parameter is also passed to the precict method, other parameters to be
added later

Commit:
Merge branch 'master' into dev_christoph

Commit:
Prepare config for real run

Edit config settings for first test. Compare parameters with
config/pretrained_model/AT_NAKO/mrge_final_40.yaml (finished running
version with NAKO_AT) as a guideline

Commit:
Filter patches with only padding

Add a functionality in keras_data_generator.py that filters out patches
with only zeros in them from the id_lists created when writing the data
to hdf5

Commit:
Make predictions in query at once for performance

Commit:
Save utility values to pickle file during exp.

Commit:
Use DataGenerator for val_data in fit of AL

Commit:
Use DataGenerator for val + print info + bugfixes

Add print commands that output info about al training loop
Use DataGenerator object for

Commit:
Train on initial data before AL loop

Add a function that chooses random patches from train data as initial
training data, add the training of initial data in CustomActiveLearner
Also change shuffle parameter and other minor changes

Commit:
Redesign init training + workers

Define num of workers in config

Train init data in __init__ method of CustomActiveLearner

Commit:
Set parameters for Exp1 + include timer

Set parameters and look over code for first experiment
Include a timer that prints the time the whole loop took as info for
possible adjustments if experiment takes too long or is very quick

Commit:
Fix utility saving + assert enough patches

create the file to save utilities if it doesn't exist
assert before starting training that there are enough patches for the
iterations

Commit:
Split query predict to not overload memory

Split the dataset of patches used in query into multiple pieces, so that
 query prediction is not performed on too much data at once

Commit:
Fix callback bug + Parameter adjustments (Exp2)

Fix callback as argument in fit so it actually works

Commit:
Add random sampling and save histories

Commit:
Fix callbacks for multiple fits + random patches

Make new callback for every use of fit to avoid overwriting
Make use of random patch selection

Commit:
Testversion for Test1

for Troubleshooting small adaptions to code

Commit:
Testversion for Test2

for Troubleshooting small adaptions to code, go back to version with
random sampling

Commit:
Test 3 - parameter changes

Commit:
Change epoch kwargs for fit -> only one callback

Commit:
Add possibility to fit an all labeled data

Add a option for the active learner to fit the model on all labeled
data, not only on the newly labeled data

Commit:
Experiment6 - Change init/train data prop for test

Commit:
New branch for troubleshooting

Commit:
Add steps_per_epoch to keras.sequence DataGen.

Commit:
Settings Exp7-2

Commit:
Partially merge test-0121 branch + end test mode

By hand merge some changes from test-0121 branch (commit d980c3f1),
mainly steps_per_epoch options for al pipeline
Revert temporary changes (comments) so that al train process can again
be used with AL

Commit:
Add comments and docstrings + change callbacks

Commit:
Use other callbacks (add. saver + ModelCheckpoint)

Commit:
Predict a single tfrecords image + Exp7-8 settings

Commit:
Exp7-9 settings

Commit:
Add option to predict only 1 img + Settings Exp8-1

Commit:
Save info of patches used in training

Save a dict including the info which hdf5 file was used (patches data)
and the IDs of the patches that were used in training + the complete
sets of init, val and train IDsk

Commit:
Save info of patches when creating hdf5 file

Save which original picture the patch belonged to (path to label and
and image + a unique id) and the patch's position

Commit:
Save the patches info as Pandas Dataframe

Commit:
Partially merge changes from dev_christoph

Get changes from dev_christoph branch (commit f0ca9f80) to test branch.
Includes mainly options to save the used patches and their origin for
later analysis

Commit:
Changes for Exp9

Move parameters for determining the used number of patches to config (as
 temporary parameters)
 Also settings for first Exp9-1

Commit:
Settings Exp9-2

Commit:
Settings Exp9-3

Commit:
Settings Exp9-4

Commit:
Settings Exp9-5

Commit:
Settings Exp9-6

Commit:
Settings Exp9-7

Commit:
Settings Exp9-8

Commit:
Settings Exp9-9

Commit:
Delete "unused" hdf5 patches data+Settings Exp8-2

Commit:
Settings Exp8-3

Commit:
Save Predictions with Exp name + Settings Predct.

Change the filename in which mosaic plots are saved so that it includes
the name of the experiment (according to config) so that the predictions
 of different experiments can be distinguished

Commit:
Save plot by slice with exp_name in filename

Also Settings for Plotting of Exp9 slices (Predict_slices_Exp9)

Commit:
Analyse saved patches info - accumulate data

Commit:
patches info - add prints + fix bin key error

Commit:
Settings For Plotting of Exp9 slices with new img

(Predict_slices_Exp9)

Commit:
Merge branch 'test-0121' into dev_christoph

Commit:
End version AL - without temporary test code

Small corrections to the results of merging the test branch with the
developement branch
Result is the momentary version of the AL pipeline at the end of my
bachelors thesis.

Commit:
Remove parts before merge with master

Commit:
Compare with master (last merge) - assimilate

Compare with the last master commit that was merged into this branch and
 match them as much as possible, again removing unnecessary changes

Commit:
Merge branch 'master' into dev_christoph

CDStark commented:
Notes: 
* Also contains changes to argparse in main.py - will likely slightly change the interaction with the code in command line
(e.g. adding flag --train will result in args.train=True, not specifying --train will result in args.train=False)
* My IDE (Pycharm) detects two errors in predict.py line 183 (indice_list_model unknown) and line 513 (read_mat unknown) in the curent master version (which I also merged into this commit
* In pipeline.py I use the max_patch_num parameter in get_fixed_patches_index() and random_shift_patch in get_patches_data()

thomaskuestner commented on med_io/pipeline.py:
```diff
@@ -38,7 +38,8 @@ def pipeline(config, dataset_image_path, dataset_label_path, dataset=None):
                                               start=config['patch_start'],
-                                              end=config['patch_end'])
+                                              end=config['patch_end'],
+                                              max_patch_num=config['max_patch_num'])
```
add to util.convert_yaml_config() check if new key max_patch_num exists, otherwise add for backward compat:
```
if not 'max_patch_num' in config.keys():
    config['max_patch_num'] = 200
```
What is a good init number? Will this affect previous pipeline?

CDStark replied:
Added the config check with the default value **None** because this is the default value of the get_fixed_patches_index function, the behaviour of the pipeline should therefore be the same for the config files which do not have this parameter
(util.convert_yaml_config() line 35)

thomaskuestner replied:
Thanks 👍

thomaskuestner commented on med_io/pipeline.py:
```diff
@@ -89,14 +90,19 @@ def _map(*args):
+                                                                      slice_channel_img=input_slice,
                                                                       slice_channel_label=output_slice,
                                                                       output_patch_size=config['model_output_size'],
+                                                                      random_shift_patch=config['random_shift_patch'],
```
add to config check

CDStark replied:
The parameter random_shift_patch already exists in the default config, is a config check necessary in that case?
If yes I would suggest using random_shift_patch=True as standard, because this is also the value in the default config and the default value of the get_patches_data function

thomaskuestner replied:
sorry, my oversight. This is not needed then.

thomaskuestner commented on med_io/pipeline.py:
```diff
@@ -89,14 +90,19 @@ def _map(*args):
             # List regularize
             #index_list = index_list / (np.array(max_data_size) + 1e-16)
-            if config['feed_pos']:
+            if config['active_learning']:
```
add to config check, default: False

CDStark replied:
Added the required config check (util.convert_yaml_config() line 37)

thomaskuestner replied:
thanks

thomaskuestner reviewed (changes requested):
add new provided config parameters to config check for backward compatibility

Commit:
Add config checks for backwards compatability

New config parameters active_learning and  max_patch_num might not be
present in some config files, therefore these checks are needed

thomaskuestner commented:
@all-contributors please add @CDStark for code, maintenance
